pkg/sql/vecindex/vecstore/store_txn.go

```diff
@@ -309,10 +333,128 @@ func (tx *Txn) SearchPartitions(
+	tableDesc catalog.TableDescriptor,
+	indexDesc catalog.Index,
+	sourceIndex catalog.Index,
+) error {
+	vectorColID := indexDesc.VectorColumnID()
+	if err := rowenc.InitIndexFetchSpec(
+		&spec.FetchSpec,
+		evalCtx.Codec,
+		tableDesc,
+		sourceIndex,
+		[]descpb.ColumnID{vectorColID},
+	); err != nil {
+		return err
+	}
+
+	vectorCol, err := catalog.MustFindColumnByID(tableDesc, vectorColID)
+	if err != nil {
+		return err
+	}
+
+	var neededColOrdinals intsets.Fast
+	neededColOrdinals.Add(vectorCol.Ordinal())
+	splitter := span.MakeSplitter(tableDesc, tableDesc.GetPrimaryIndex(), neededColOrdinals)
+	spec.FamilyIDs = splitter.FamilyIDs()
+
+	// We need to decode the columns for the PK from the vector index. If the
+	// vector index is being mutated, there's a chance the primary key is changing.
+	// The vector index needs to point into the NEW primary key, so find that here.
+	var primaryKey catalog.Index
+	primaryKey = tableDesc.GetPrimaryIndex()
+	if indexDesc.IsMutation() {
+		for _, idx := range tableDesc.NonPrimaryIndexes() {
+			if idx.GetEncodingType() == catenumpb.PrimaryIndexEncoding && !idx.IsTemporaryIndexForBackfill() {
+				primaryKey = idx
+				break
+			}
+		}
+	}
+	keyCols := primaryKey.CollectKeyColumnIDs().Ordered()
+	return rowenc.InitIndexFetchSpec(&spec.ExtractPKFetchSpec, evalCtx.Codec, tableDesc, indexDesc, keyCols)
+}
+
+// PKDecoder is used to extract the primary key from a vector index.
+type pkDecoder struct {
+	fetchSpec *fetchpb.IndexFetchSpec
+	output    rowenc.EncDatumRow
+	scratch   rowenc.EncDatumRow
+	colOrdMap catalog.TableColMap
+}
+
+// Init initializes the PKDecoder with the given fetch spec from the
+// GetFullVectorsFetchSpec.
+func (d *pkDecoder) Init(fetchSpec *fetchpb.IndexFetchSpec) {
+	d.fetchSpec = fetchSpec
+	if cap(d.output) < len(fetchSpec.FetchedColumns) {
+		d.output = make(rowenc.EncDatumRow, len(fetchSpec.FetchedColumns))
+	} else {
+		d.output = d.output[:0]
+	}
+	for i, col := range fetchSpec.FetchedColumns {
+		d.colOrdMap.Set(col.ColumnID, i)
+	}
+}
+
+// Decode decodes the primary key from the given key bytes. The key returned
+// remains valid until the next row is decoded.
+func (d *pkDecoder) Decode(
+	treeKey cspann.TreeKey, keyBytes []byte,
+) (row rowenc.EncDatumRow, err error) {
+	decodeFromKey := func(keyCols []fetchpb.IndexFetchSpec_KeyColumn, keyBytes []byte) error {
+		if len(keyCols) == 0 {
+			if len(keyBytes) > 0 {
+				return errors.AssertionFailedf("expected empty key bytes")
+			}
+			return nil
+		}
+		if cap(d.scratch) < len(keyCols) {
+			d.scratch = make(rowenc.EncDatumRow, len(keyCols))
+		}
+		d.scratch = d.scratch[:len(keyCols)]
+		_, _, err = rowenc.DecodeKeyValsUsingSpec(keyCols, keyBytes, d.scratch)
+		if err != nil {
+			return err
+		}
+		for i, col := range keyCols {
+			idx, ok := d.colOrdMap.Get(col.ColumnID)
+			if !ok {
+				continue
+			}
+			d.output[idx] = d.scratch[i]
+		}
+		return nil
+	}
+	// Decode the index prefix columns, if any,
```

Comment:
nit: comma at the end.

Also, maybe mention that we only need to do this if PK columns are included in the prefix.